src/v/ssx/process.cc

```diff
@@ -0,0 +1,180 @@
+process::exit_status_t
+unwrap_wait_status(ss::experimental::process::wait_status& result) {
+    return ss::visit(
+      result,
+      [](ss::experimental::process::wait_exited exited) {
+          return process::exit_status_t{
+            .exit_int = exited.exit_code, .exit_reason = "exit code"};
+      },
+      [](ss::experimental::process::wait_signaled signaled) {
+          return process::exit_status_t{
+            .exit_int = signaled.terminating_signal,
+            .exit_reason = "exit signal"};
+      });
+}
+} // namespace
+
+process::~process() {
+    _as.request_abort();
+    vassert(
+      !is_running(),
+      "Processes must exit before destruction: cmd {}",
+      _cmd_str);
+}
+
+ss::future<std::error_code> process::spawn(
+  const std::filesystem::path& cmd, ss::experimental::spawn_parameters params) {
+    gate_guard g{_gate};
+    if (is_running()) {
+        vlog(proclog.error, "A process is already running: cmd {}", _cmd_str);
+        co_return process_errc::running;
+    }
+
+    // Set command string for logging purposes
+    _cmd_str = cmd.string();
+
+    auto p = co_await ss::experimental::spawn_process(cmd, std::move(params));
+    _process.emplace(std::move(p));
+
+    // Capture output async in the background so the broker is not blocked.
+    ssx::background = ssx::spawn_with_gate_then(_gate, [this] {
+        return consume_input_stream(_process->stdout(), true, _as)
+          .then([this] {
+              return consume_input_stream(_process->stderr(), false, _as);
+          })
+          // According to seastar docs, a process that is not wait'd may leave a
+          // zombie behind. So we do that here.
+          .then([this] { return handle_wait(); })
+          .finally([this] { _process.reset(); });
+    });
+
+    co_return process_errc::success;
+}
+
+ss::future<> process::stop() { co_await _gate.close(); }
+
+process::exit_status_t process::exit_status() {
+    return unwrap_wait_status(_wait_status);
+}
+
+ss::future<> process::handle_wait() {
+    vassert(is_running(), "_process not instantiated");
+
+    _wait_status = co_await _process->wait();
+    auto status = exit_status();
+
+    // There is no signal=0, so an exit_int=0 is the success exit code.
+    if (status.exit_int != 0) {
+        vlog(
+          proclog.error,
+          "Process stop fail: cmd {}, {}={}",
+          _cmd_str,
+          status.exit_reason,
+          status.exit_int);
+    } else {
+        vlog(proclog.trace, "Process stop success: cmd {}", _cmd_str);
+    }
+}
+
+ss::future<std::error_code>
+process::terminate(std::chrono::milliseconds timeout) {
+    gate_guard g{_gate};
+    if (!is_running()) {
+        co_return process_errc::does_not_exist;
+    }
+
+    try {
+        _process->terminate();
+    } catch (const std::system_error&) {
+        // The process is already dead
+        co_return process_errc::does_not_exist;
+    }
+
+    auto deadline = ss::steady_clock_type::now() + timeout;
```

Comment:
nit: generally with timeouts we use the lowres clock in seastar as far as I can tell.